### PR TITLE
Make OSSL_CORE_BIO distinct from and not castable to BIO

### PR DESCRIPTION
**Pull request by paulidale**
Refer to #14418 for a description of the problem being solved. 

- [ ] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

**Comment by paulidale** [comment]
Bring this out of WIP.  I don't think the test failures are relevant.

File: include/openssl/core_dispatch.h [hunk 1]
```diff
@@ -135,12 +135,13 @@ OSSL_CORE_MAKE_FUNC(void,
 #define OSSL_FUNC_BIO_NEW_MEMBUF              41
 #define OSSL_FUNC_BIO_READ_EX                 42
 #define OSSL_FUNC_BIO_WRITE_EX                43
-#define OSSL_FUNC_BIO_FREE                    44
-#define OSSL_FUNC_BIO_VPRINTF                 45
-#define OSSL_FUNC_BIO_VSNPRINTF               46
-#define OSSL_FUNC_BIO_PUTS                    47
-#define OSSL_FUNC_BIO_GETS                    48
-#define OSSL_FUNC_BIO_CTRL                    49
+#define OSSL_FUNC_BIO_UP_REF                  44
```
Comment (t8m) [inline include/openssl/core_dispatch.h:138]: This should be added to the provider-base.pod documentation.

File: test/ossl_store_test.c [hunk 2]
```diff
@@ -131,6 +138,7 @@ static int test_store_get_params(int idx)
 
     return 1;
 }
+#include <unistd.h>
```
Comment (t8m) [inline test/ossl_store_test.c:141]: This looks weirdly placed. Why not to have it at the top?

**Review by t8m — CHANGES_REQUESTED** [review]

File: crypto/store/store_lib.c [hunk 3]
```diff
@@ -941,9 +942,10 @@ OSSL_STORE_CTX *OSSL_STORE_attach(BIO *bp, const char *scheme,
         const OSSL_PROVIDER *provider =
             OSSL_STORE_LOADER_provider(fetched_loader);
         void *provctx = OSSL_PROVIDER_get0_provider_ctx(provider);
+        OSSL_CORE_BIO *cbio = ossl_core_bio_new_from_bio(bp);
 
-        if ((loader_ctx =
-             fetched_loader->p_attach(provctx, (OSSL_CORE_BIO *)bp)) == NULL) {
+        if (cbio != NULL
+            && (loader_ctx = fetched_loader->p_attach(provctx, cbio)) == NULL) {
```
[the remark this answers is not in the document]
Comment (paulidale) [reply crypto/store/store_lib.c:948]: Good catch.  I'll fix.

[the remark this answers is not in the document]
Comment (paulidale) [reply crypto/store/store_lib.c:948]: Just not handling the failures properly.

File: test/ossl_store_test.c (hunk 2 above)
Comment (paulidale) [reply test/ossl_store_test.c:141]: Debugging leftover trying to figure out why the file wasn't found...

File: include/openssl/core_dispatch.h (hunk 1 above)
Comment (paulidale) [reply include/openssl/core_dispatch.h:138]: Will do.

Comment (paulidale) [reply include/openssl/core_dispatch.h:138]: There are a number of other issues with the provider-base documentations which I've also fixed.

**Comment by paulidale** [comment]
Comments addressed.

**Review by t8m — APPROVED** [review]

**Comment by levitte** [comment]
I would like to see the story told better in the commit messages.  Right now, the messages are *very* short and don't really tell the story much at all to an external onlooker.

**Comment by paulidale** [comment]
I'll add some words when squishing things up a bit.

**Comment by paulidale** [comment]
Merged to master with the main commit message:
Add a real type for OSSL_CORE_BIO which is distinct from and not castable to BIO
    
> Providers (particularly the FIPS provider) needs access to BIOs from libcrypto.
> Libcrypto is allowed to change the internal format of the BIO structure and it
> is still expected to work with providers that were already built.  This means
> that the libcrypto BIO must be distinct from and not castable to the provider
> side OSSL_CORE_BIO.

**Pull request closed without merging**